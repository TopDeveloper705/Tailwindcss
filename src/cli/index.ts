import arg from 'arg';
import { resolve } from 'path';
import { Processor } from '../lib';
import { readFileSync, writeFileSync } from 'fs';
import { HTMLParser } from '../utils/parser';
import { StyleSheet } from '../utils/style';
import {
  getVersion,
  FilePattern,
  walk,
  isFile,
  generateTemplate,
  Console,
} from './utils';

const doc = `
Generate css from text files that containing tailwindcss classes.
By default, it will use interpretation mode to generate a single css file.

Usage:
  windicss [filenames]
  windicss [filenames] -c -m
  windicss [filenames] -c -s -m
  windicss [filenames] [-c | -i] [-b | -s] [-m] [-p <prefix:string>] [-o <path:string>] [--args arguments]

Options:
  -h, --help            Print this help message and exit.
  -v, --version         Print windicss current version and exit.

  -i, --interpret       Interpretation mode, generate class name corresponding to tailwindcss. This is the default behavior.
  -c, --compile         Compilation mode, combine the class name in each row into a single class.
  -t, --preflight       Add preflights, default is false.

  -b, --combine         Combine all css into one single file. This is the default behavior.
  -s, --separate        Generate a separate css file for each input file.

  -m, --minify          Generate minimized css file.
  -p, --prefix PREFIX   Set the css class name prefix, only valid in compilation mode. The default prefix is 'windi-'.
  -o, --output PATH     Set output css file path.
  -f, --config PATH     Set config file path.

  --init PATH           Start a new project on the path.
`;

const args = arg({
  // Types
  '--help': Boolean,
  '--version': Boolean,
  '--compile': Boolean,
  '--interpret': Boolean,
  '--preflight': Boolean,
  '--combine': Boolean,
  '--separate': Boolean,
  '--minify': Boolean,
  '--init': String,
  '--prefix': String,
  '--output': String,
  '--config': String,

  // Aliases
  '-h': '--help',
  '-v': '--version',
  '-i': '--interpret',
  '-c': '--compile',
  '-t': '--preflight',
  '-b': '--combine',
  '-s': '--separate',

  '-m': '--minify',
  '-p': '--prefix',
  '-o': '--output',
  '-f': '--config',
});

if (args['--help'] || (args._.length === 0 && Object.keys(args).length === 1)) {
  Console.log(doc);
  process.exit();
}

if (args['--version']) {
  Console.log(getVersion());
  process.exit();
}

if (args['--init']) {
  const template = generateTemplate(args['--init'], args['--output']);
  args._.push(template.html);
  args['--preflight'] = true;
  args['--output'] = template.css;
}

const localFiles = walk('.', true).filter((i) => i.type === 'file').map(i=> ({ type: i.type, path: i.path.replace(/\\/g, '/') }) );

const matchFiles: string[] = [];
for (const pt of args._) {
  if (isFile(pt) && pt.search(/\.windi\./) === -1) {
    matchFiles.push(pt);
  } else if (pt.search(/\*/) !== -1) {
    // match files like **/*.html **/src/*.html *.html ...
    localFiles.forEach((i) => {
      const pattern = new FilePattern(pt);
      if (pattern.match(i.path) && i.path.search(/\.windi\./) === -1)
        matchFiles.push(i.path);
    });
  } else {
    Console.error(`File ${pt} does not exist!`);
  }
}

let ignoredClasses: string[] = [];
const preflights: StyleSheet[] = [];
const styleSheets: StyleSheet[] = [];
const processor = new Processor(args['--config'] ? require(resolve(args['--config'])) : undefined);

if (args['--compile']) {
  // compilation mode
  const prefix = args['--prefix'] ?? 'windi-';
  matchFiles.forEach((file) => {
    let indexStart = 0;
    const outputStyle: StyleSheet[] = [];
    const outputHTML: string[] = [];
    const html = readFileSync(file).toString();
    const parser = new HTMLParser(html);

    // Match tailwind ClassName then replace with new ClassName
    parser.parseClasses().forEach((p) => {
      outputHTML.push(html.substring(indexStart, p.start));
      const utility = processor.compile(p.result, prefix, true); // Set third argument to false to hide comments;
      outputStyle.push(utility.styleSheet);
      ignoredClasses = [...ignoredClasses, ...utility.ignored];
      outputHTML.push([utility.className, ...utility.ignored].join(' '));
      indexStart = p.end;
    });
    outputHTML.push(html.substring(indexStart));
    styleSheets.push(
      outputStyle.reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
    );

    const outputFile = file.replace(/(?=\.\w+$)/, '.windi');
    writeFileSync(outputFile, outputHTML.join(''));
    Console.log(`${file} -> ${outputFile}`);

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
} else {
  // interpretation mode
  matchFiles.forEach((file) => {
    const parser = new HTMLParser(readFileSync(file).toString());
    const utility = processor.interpret(
      parser
        .parseClasses()
        .map((i) => i.result)
        .join(' ')
    );
    styleSheets.push(utility.styleSheet);
    ignoredClasses = [...ignoredClasses, ...utility.ignored];

    if (args['--preflight'])
      preflights.push(processor.preflight(parser.html));
  });
}

if (args['--separate']) {
  styleSheets.forEach((style, index) => {
    const filePath = matchFiles[index].replace(/\.\w+$/, '.windi.css');
    if (args['--preflight']) style = preflights[index].extend(style);
    writeFileSync(filePath, style.build(args['--minify']));
    Console.log(`${matchFiles[index]} -> ${filePath}`);
  });
} else {
  let outputStyle = styleSheets
    .reduce(
      (previousValue: StyleSheet, currentValue: StyleSheet) =>
        previousValue.extend(currentValue),
      new StyleSheet()
    )
    .combine()
    .sort();
  if (args['--preflight'])
    outputStyle = preflights
      .reduce(
        (previousValue: StyleSheet, currentValue: StyleSheet) =>
          previousValue.extend(currentValue),
        new StyleSheet()
      )
      .combine()
      .sort()
      .extend(outputStyle);
  const filePath = args['--output'] ?? 'windi.output.css';
  writeFileSync(filePath, outputStyle.build(args['--minify']));
  Console.log('matched files:', matchFiles);
  Console.log('output file:', filePath);
}

Console.log('ignored classes:', ignoredClasses);
