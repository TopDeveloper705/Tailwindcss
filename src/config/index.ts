export { baseConfig } from './base';
export { colors } from './colors';
export const twExclude = [
  /-hex-/,                          // disable hex color
  /-\$/,                            // disable variable
  /(rem|em|px|rpx|vh|vw|ch|ex)$/,   // disable size
  /-\d*[13579]$/,                   // disable odd number
  /-([0-9]{1,}[.][0-9]*)$/,         // disable float
  /^!/,                             // disable important utility
];
