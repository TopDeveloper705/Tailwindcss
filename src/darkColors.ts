export { darkColors as default } from './config';
