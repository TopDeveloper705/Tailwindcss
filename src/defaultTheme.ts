import { Processor } from "./lib";

export default new Processor().allTheme;
