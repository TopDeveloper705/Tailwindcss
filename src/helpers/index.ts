export * from './template';
export * from './transform';
