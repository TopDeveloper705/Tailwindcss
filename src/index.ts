export * from './processor';
export * from './utils';
export * from './config';
export * from './plugin';

module.exports = {
    'colors': require('./config/colors')
}

