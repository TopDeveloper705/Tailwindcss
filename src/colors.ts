import { default as colors } from './config/colors';

module.exports = colors;