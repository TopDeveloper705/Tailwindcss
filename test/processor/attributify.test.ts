import { Processor } from '../../src/lib';

const processor = new Processor();

describe('Attributify Mode', () => {
  it('simple attributify', () => {
    expect(processor.attributify({
      font: 'bold',
      p: ['x-4', 'y-2', 'lg:4'],
      bg: ['green-400', 'opacity-50'],
    }).styleSheet.build()).toMatchSnapshot('css');
  });

  it('with variants', () => {
    expect(processor.attributify({
      sm: ['bg-red-500', 'bg-opacity-50'],
      'sm:hover': ['text-red-500', 'text-lg'],
      'sm-hover': ['text-red-500', 'text-lg'],
    }).styleSheet.build()).toMatchSnapshot('css');
  });

  it('mix utility and variant', () => {
    expect(processor.attributify({
      p: ['x-4', 'y-2', 'md:x-2', 'lg:hover:x-8'],
      sm: ['bg-red-500', 'bg-opacity-50', 'hover:bg-green-300'],
      'sm:hover': ['text-red-500', 'text-lg', 'focus:bg-white'],
    }).styleSheet.build()).toMatchSnapshot('css');
  });

  it('with variants and utility', () => {
    expect(processor.attributify({
      'sm:p': ['x-4', 'y-2'],
      'hover-text': ['red-300', 'sm'],
      'hover:text': ['red-500', 'lg'],
      'sm:hover:text': ['red-500', 'lg'],
      'sm-hover-text': ['red-300', 'sm'],
    }).styleSheet.build()).toMatchSnapshot('css');
  });

  it('with negative utility', () => {
    expect(processor.attributify({
      'm': ['-x-4', 'md:y-2'],
      'sm': ['-my-2'],
    }).styleSheet.build()).toMatchSnapshot('css');
  });

  it('with grid utility', () => {
    const result = processor.attributify({
      'grid': [
        'default', 'inline', // display
        'cols-1', 'cols-none', // grid-template-columns
        'col-auto', 'col-span-2', // grid-column
        'rows-3', 'rows-none', // grid-template-rows
        'row-auto', 'row-span-2', // grid-rows
        'flow-row', 'flow-col', 'flow-row-dense', // grid-auto-flow
        'auto-cols-auto', 'auto-cols-min', 'auto-cols-max', // grid-auto-columns
        'auto-rows-auto', 'auto-rows-min', 'auto-rows-max', // grid-auto-rows
        'gap-2', 'gap-x-4', 'gap-y-2', // gap/column-gap/row-gap
      ],
    });
    expect(result.ignored.length).toEqual(0);
    expect(result.styleSheet.build()).toMatchSnapshot('css');
  });

  it('with align utility', () => {
    const result = processor.attributify({
      'align': [
        'content-center', 'content-end', // align-content
        'items-start', 'items-center', // align-items
        'self-auto', 'self-end', // align-self
      ],
    });
    expect(result.ignored.length).toEqual(0);
    expect(result.styleSheet.build()).toMatchSnapshot('css');
  });
});
