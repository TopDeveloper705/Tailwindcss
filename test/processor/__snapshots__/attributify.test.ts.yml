Tools / mix utility and variant / css / 0: |-
  [p~="x-4"] {
    padding-left: 1rem;
    padding-right: 1rem;
  }
  [p~="y-2"] {
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
  }
  @media (min-width: 640px) {
    [sm~="bg-red-500"] {
      --tw-bg-opacity: 1;
      background-color: rgba(239, 68, 68, var(--tw-bg-opacity));
    }
    [sm~="hover:bg-green-300"]:hover {
      --tw-bg-opacity: 1;
      background-color: rgba(110, 231, 183, var(--tw-bg-opacity));
    }
    [sm\:hover~="focus:bg-white"]:hover:focus {
      --tw-bg-opacity: 1;
      background-color: rgba(255, 255, 255, var(--tw-bg-opacity));
    }
    [sm~="bg-opacity-50"] {
      --tw-bg-opacity: 0.5;
    }
    [sm\:hover~="text-lg"]:hover {
      font-size: 1.125rem;
      line-height: 1.75rem;
    }
    [sm\:hover~="text-red-500"]:hover {
      --tw-text-opacity: 1;
      color: rgba(239, 68, 68, var(--tw-text-opacity));
    }
  }
  @media (min-width: 768px) {
    [p~="md:x-2"] {
      padding-left: 0.5rem;
      padding-right: 0.5rem;
    }
  }
  @media (min-width: 1024px) {
    [p~="lg:hover:x-8"]:hover {
      padding-left: 2rem;
      padding-right: 2rem;
    }
  }
Tools / simple attributify / css / 0: |-
  [bg~="green-400"] {
    --tw-bg-opacity: 1;
    background-color: rgba(52, 211, 153, var(--tw-bg-opacity));
  }
  [bg~="opacity-50"] {
    --tw-bg-opacity: 0.5;
  }
  [font="bold"] {
    font-weight: 700;
  }
  [p~="x-4"] {
    padding-left: 1rem;
    padding-right: 1rem;
  }
  [p~="y-2"] {
    padding-top: 0.5rem;
    padding-bottom: 0.5rem;
  }
  @media (min-width: 1024px) {
    [p~="lg:4"] {
      padding: 1rem;
    }
  }
Tools / with align utility / css / 0: |-
  [align~="items-start"] {
    -webkit-box-align: start;
    -ms-flex-align: start;
    -webkit-align-items: flex-start;
    align-items: flex-start;
  }
  [align~="items-center"] {
    -webkit-box-align: center;
    -ms-flex-align: center;
    -webkit-align-items: center;
    align-items: center;
  }
  [align~="content-center"] {
    -ms-flex-line-pack: center;
    -webkit-align-content: center;
    align-content: center;
  }
  [align~="content-end"] {
    -ms-flex-line-pack: end;
    -webkit-align-content: flex-end;
    align-content: flex-end;
  }
  [align~="self-auto"] {
    -ms-flex-item-align: auto;
    -ms-grid-row-align: auto;
    -webkit-align-self: auto;
    align-self: auto;
  }
  [align~="self-end"] {
    -ms-flex-item-align: end;
    -webkit-align-self: flex-end;
    align-self: flex-end;
  }
Tools / with box utility / css / 0: |-
  [box~="decoration-slice"] {
    -webkit-box-decoration-break: slice;
    box-decoration-break: slice;
  }
  [box~="decoration-clone"] {
    -webkit-box-decoration-break: clone;
    box-decoration-break: clone;
  }
  [box~="border"] {
    -webkit-box-sizing: border-box;
    box-sizing: border-box;
  }
  [box~="content"] {
    -webkit-box-sizing: content-box;
    box-sizing: content-box;
  }
  [box~="shadow"] {
    --tw-shadow-color: 0, 0, 0;
    --tw-shadow: 0 1px 3px 0 rgba(var(--tw-shadow-color), 0.1), 0 1px 2px 0 rgba(var(--tw-shadow-color), 0.06);
    -webkit-box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);
    box-shadow: var(--tw-ring-offset-shadow, 0 0 #0000), var(--tw-ring-shadow, 0 0 #0000), var(--tw-shadow);
  }
  [box~="shadow-gray-200"] {
    --tw-shadow-color: 229, 231, 235;
  }
Tools / with display utility / css / 0: |-
  [display~="block"] {
    display: block;
  }
  [display~="inline-block"] {
    display: inline-block;
  }
  [display~="inline"] {
    display: inline;
  }
  [display~="flow-root"] {
    display: flow-root;
  }
  [display~="contents"] {
    display: contents;
  }
  [display~="list-item"] {
    display: list-item;
  }
  [display~="hidden"] {
    display: none;
  }
  [display~="visible"] {
    visibility: visible;
  }
  [display~="invisible"] {
    visibility: hidden;
  }
  @media (min-width: 768px) {
    [display~="md:hidden"] {
      display: none;
    }
  }
Tools / with gradient utility / css / 0: |-
  [gradient~="none"] {
    background-image: none;
  }
  [gradient~="to-r"] {
    background-image: -o-linear-gradient(left, var(--tw-gradient-stops));
    background-image: -webkit-gradient(linear, left top, right top, from(var(--tw-gradient-stops)));
    background-image: linear-gradient(to right, var(--tw-gradient-stops));
  }
  [gradient~="to-br"] {
    background-image: -o-linear-gradient(top left, var(--tw-gradient-stops));
    background-image: -webkit-gradient(linear, left top, right bottom, from(var(--tw-gradient-stops)));
    background-image: linear-gradient(to bottom right, var(--tw-gradient-stops));
  }
  [gradient~="from-yellow-400"] {
    --tw-gradient-from: #fbbf24;
    --tw-gradient-stops: var(--tw-gradient-from), var(--tw-gradient-to, rgba(251, 191, 36, 0));
  }
  [gradient~="via-red-500"] {
    --tw-gradient-stops: var(--tw-gradient-from), #ef4444, var(--tw-gradient-to, rgba(239, 68, 68, 0));
  }
  [gradient~="to-pink-500"] {
    --tw-gradient-to: #ec4899;
  }
Tools / with grid utility / css / 0: |-
  [grid~="default"] {
    display: -ms-grid;
    display: grid;
  }
  [grid~="inline"] {
    display: -ms-inline-grid;
    display: inline-grid;
  }
  [grid~="gap-2"] {
    grid-gap: 0.5rem;
    gap: 0.5rem;
  }
  [grid~="gap-x-4"] {
    -webkit-column-gap: 1rem;
    -moz-column-gap: 1rem;
    grid-column-gap: 1rem;
    column-gap: 1rem;
  }
  [grid~="gap-y-2"] {
    -webkit-row-gap: 0.5rem;
    -moz-row-gap: 0.5rem;
    grid-row-gap: 0.5rem;
    row-gap: 0.5rem;
  }
  [grid~="flow-row"] {
    grid-auto-flow: row;
  }
  [grid~="flow-col"] {
    grid-auto-flow: column;
  }
  [grid~="flow-row-dense"] {
    grid-auto-flow: row dense;
  }
  [grid~="cols-1"] {
    grid-template-columns: repeat(1, minmax(0, 1fr));
  }
  [grid~="cols-none"] {
    grid-template-columns: none;
  }
  [grid~="auto-cols-auto"] {
    grid-auto-columns: auto;
  }
  [grid~="auto-cols-min"] {
    grid-auto-columns: -webkit-min-content;
    grid-auto-columns: min-content;
  }
  [grid~="auto-cols-max"] {
    grid-auto-columns: -webkit-max-content;
    grid-auto-columns: max-content;
  }
  [grid~="col-auto"] {
    grid-column: auto;
  }
  [grid~="col-span-2"] {
    -ms-grid-column-span: span 2 / span 2;
    grid-column: span 2 / span 2;
  }
  [grid~="rows-3"] {
    grid-template-rows: repeat(3, minmax(0, 1fr));
  }
  [grid~="rows-none"] {
    grid-template-rows: none;
  }
  [grid~="auto-rows-auto"] {
    grid-auto-rows: auto;
  }
  [grid~="auto-rows-min"] {
    grid-auto-rows: -webkit-min-content;
    grid-auto-rows: min-content;
  }
  [grid~="auto-rows-max"] {
    grid-auto-rows: -webkit-max-content;
    grid-auto-rows: max-content;
  }
  [grid~="row-auto"] {
    grid-row: auto;
  }
  [grid~="row-span-2"] {
    -ms-grid-row-span: span 2 / span 2;
    grid-row: span 2 / span 2;
  }
Tools / with negative utility / css / 0: |-
  [m~="-x-4"] {
    margin-left: -1rem;
    margin-right: -1rem;
  }
  @media (min-width: 640px) {
    [sm~="-my-2"] {
      margin-top: -0.5rem;
      margin-bottom: -0.5rem;
    }
  }
  @media (min-width: 768px) {
    [m~="md:y-2"] {
      margin-top: 0.5rem;
      margin-bottom: 0.5rem;
    }
  }
Tools / with position utility / css / 0: |-
  [position~="order-1"] {
    -webkit-box-ordinal-group: 2;
    -webkit-order: 1;
    -ms-flex-order: 1;
    order: 1;
  }
  [position~="order-first"] {
    -webkit-box-ordinal-group: -9998;
    -webkit-order: -9999;
    -ms-flex-order: -9999;
    order: -9999;
  }
  [position~="float-right"] {
    float: right;
  }
  [position~="float-left"] {
    float: left;
  }
  [position~="float-none"] {
    float: none;
  }
  [position~="clear-left"] {
    clear: left;
  }
  [position~="clear-right"] {
    clear: right;
  }
  [position~="clear-both"] {
    clear: both;
  }
  [position~="clear-none"] {
    clear: none;
  }
  [position~="static"] {
    position: static;
  }
  [position~="fixed"] {
    position: fixed;
  }
  [position~="absolute"] {
    position: absolute;
  }
  [position~="relative"] {
    position: relative;
  }
  [position~="sticky"] {
    position: sticky;
    position: -webkit-sticky;
  }
  [position~="inset-1"] {
    top: 0.25rem;
    right: 0.25rem;
    bottom: 0.25rem;
    left: 0.25rem;
  }
  [position~="-inset-1"] {
    top: -0.25rem;
    right: -0.25rem;
    bottom: -0.25rem;
    left: -0.25rem;
  }
  [position~="-inset-y-2"] {
    top: -0.5rem;
    bottom: -0.5rem;
  }
  [position~="inset-x-1"] {
    right: 0.25rem;
    left: 0.25rem;
  }
  [position~="top-1.5"] {
    top: 0.375rem;
  }
  [position~="-left-3/4"] {
    left: -75%;
  }
Tools / with variants / css / 0: |-
  @media (min-width: 640px) {
    [sm~="bg-red-500"] {
      --tw-bg-opacity: 1;
      background-color: rgba(239, 68, 68, var(--tw-bg-opacity));
    }
    [sm~="bg-opacity-50"] {
      --tw-bg-opacity: 0.5;
    }
    [sm\:hover~="text-lg"]:hover {
      font-size: 1.125rem;
      line-height: 1.75rem;
    }
    [sm-hover~="text-lg"]:hover {
      font-size: 1.125rem;
      line-height: 1.75rem;
    }
    [sm\:hover~="text-red-500"]:hover {
      --tw-text-opacity: 1;
      color: rgba(239, 68, 68, var(--tw-text-opacity));
    }
    [sm-hover~="text-red-500"]:hover {
      --tw-text-opacity: 1;
      color: rgba(239, 68, 68, var(--tw-text-opacity));
    }
  }
Tools / with variants and utility / css / 0: |-
  [hover-text~="sm"]:hover {
    font-size: 0.875rem;
    line-height: 1.25rem;
  }
  [hover\:text~="lg"]:hover {
    font-size: 1.125rem;
    line-height: 1.75rem;
  }
  [hover-text~="red-300"]:hover {
    --tw-text-opacity: 1;
    color: rgba(252, 165, 165, var(--tw-text-opacity));
  }
  [hover\:text~="red-500"]:hover {
    --tw-text-opacity: 1;
    color: rgba(239, 68, 68, var(--tw-text-opacity));
  }
  @media (min-width: 640px) {
    [sm\:hover\:text~="lg"]:hover {
      font-size: 1.125rem;
      line-height: 1.75rem;
    }
    [sm-hover-text~="sm"]:hover {
      font-size: 0.875rem;
      line-height: 1.25rem;
    }
    [sm\:p~="x-4"] {
      padding-left: 1rem;
      padding-right: 1rem;
    }
    [sm\:p~="y-2"] {
      padding-top: 0.5rem;
      padding-bottom: 0.5rem;
    }
    [sm\:hover\:text~="red-500"]:hover {
      --tw-text-opacity: 1;
      color: rgba(239, 68, 68, var(--tw-text-opacity));
    }
    [sm-hover-text~="red-300"]:hover {
      --tw-text-opacity: 1;
      color: rgba(252, 165, 165, var(--tw-text-opacity));
    }
  }
